cryptoki/src/session/mod.rs

```diff
@@ -35,37 +35,37 @@ pub use flags::*;
 /// threads. A Session needs to be created in its own thread or to be passed by ownership to
 /// another thread.
 #[derive(Debug)]
-pub struct Session<'a> {
+pub struct Session {
     handle: CK_SESSION_HANDLE,
-    client: &'a Pkcs11,
+    client: Pkcs11,
     // This is not used but to prevent Session to automatically implement Send and Sync
     _guard: *mut u32,
```

Comment:
Could you get rid of this and simply do an `impl !Sync for Session {}` below?

Reply:
I think this should be part of a larger discussion.  The original purpose, afaik, for this `_guard` was to force a compile check to prevent a `Session` be used across multiple threads.  Should we maintain that for `Session` or allow a user to, in effect, do whatever they want with the Session, even if it's not in the best "pkcs11" way?

For example I started making changes to Parsec to support these changes but immediately ran into a problem where the PKCS11 provider was unable to be used within an `Arc` because I put `Session` within it to support a `management` session.

Reply:
> Should we maintain that for Session or allow a user to, in effect, do whatever they want with the Session, even if it's not in the best "pkcs11" way?

Isn't that a no-no, based on the spec? As in, multithreaded use of a session _may_ lead to inconsistencies. I was only suggesting to replace the "indirect" way of making `Session` not `Sync` with an explicit one. The broader discussion can be left for another issue.

Reply:
The spec states that:
>  the behavior of Cryptoki is _undefined_ if multiple threads of an application attempt to access a common Cryptoki session simultaneously

Is it disallowed?  It's highly suggested that you do _not_ use >1 thread to access the same session, but it's not a hard 'no'.

I like that we attempt to protect the user, I think that makes a lot of sense and I'm not against keeping it.  I think user programs (e.g. Parsec) may have a hard time using a management type session without some architecting around it.

Reply:
Oh so you can't do negative trait bounds yet:

```
error[E0658]: negative trait bounds are not yet fully implemented; use marker types for now
  --> cryptoki\src\session\mod.rs:64:6
   |
64 | impl !Sync for Session {}
   |      ^^^^^
   |
   = note: see issue #68318 <https://github.com/rust-lang/rust/issues/68318> for more information
```

Reply:
Damn, ok, thanks for trying! We can just keep it as is for now

Reply:
I think you probably want [PhantomData](https://doc.rust-lang.org/nomicon/phantom-data.html) here. Also, you can't implement negative traits (yet), but I think you can test for them to confirm you're getting the effect you want. (Note: I'm not near enough to real code to check that this is even syntactically correct, so it may require some massaging.)

```rust
#[cfg(test)]
mod test {
    use super::MyType;

    fn not_sync<T:!Sync>(_unused: T) {}

    #[test]
    fn my_type_not_sync() {
        // Won't compile if MyType is Sync
        not_sync(MyType);
    }
}
```

Reply:
That does look pretty nice. @mjb3279 - would you prefer keeping as is or using `PhantomData`? The test would also be good

Reply:
I can do that.  Been busy the last few days so haven't had a chance to integrate this idea.  Will try to get to it by the end of the week.

Reply:
No rush on my side, I just got back to this.

Reply:
Looks like you still can't use negative bounds as you described above.